### ISSUE TITLE
Error: Exception calling application: 'coroutine' object has no attribute 'outs' in dynamic provider with async CRUD methods

### ISSUE THREAD

**Issue by msamsami** [pulumi/pulumi#13153]
### What happened?

I have a dynamic provider with async CRUD methods. The reason for making them async is that I'm using another method `get_access_token` in them which calls the `pulumi_azure_native.authorization.get_client_token()` method which returns an awaitable.

```python
class AccessPolicyProvider(ResourceProvider, _AzureRESTAPI):

    async def create(self, args) -> CreateResult:
        tenant_id, subscription_id, access_token = await self.get_access_token()
        body = self.keyvault_access_policy_req_body(
            tenant_id=tenant_id,
            object_id=args['object_id'],
            certificate_permissions=args['certificate_permissions'],
            key_permissions=args['key_permissions'],
            secret_permissions=args['secret_permissions']
        )
        url = self.keyvault_access_policy_uri(subscription_id, args['resource_group_name'], args['account_name'], 'add')

        _ = self.rest_api_req(url=url, body=body, access_token=access_token, method='PUT')

        return CreateResult("keyvault-access-policy-" + binascii.b2a_hex(os.urandom(8)).decode("utf-8"), outs=args)

    async def delete(self, id, args):
        ....
    
    async def update(self, id, old_inputs, new_inputs) -> UpdateResult:
        ...
```

However, I get the following error during `pulumi up` which seems to be related to this async create method:

```
  pulumi-python:dynamic/keyvault:AccessPolicy (kv-access-policy-web-app):
    error: Exception calling application: 'coroutine' object has no attribute 'outs'

  pulumi:pulumi:Stack (platform-auto):
    D:\Infrastructure\venv\lib\site-packages\grpc\_server.py:456: RuntimeWarning: coroutine 'AccessPolicyProvider.create' was never awaited
      return None, False
    RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

### Expected Behavior

I expected that Pulumi automatically detects that these CRUD methods are async and use proper `await`-ing for calling them.

### Steps to reproduce

The issue is directly related to the dynamic provider class I mentioned above.

### Output of `pulumi about`

CLI          
Version      3.47.1
Go Version   go1.19.2
Go Compiler  gc

Plugins
NAME          VERSION
azure-native  1.67.0
python        unknown

Host
OS       Microsoft Windows 11 Home
Version  10.0.22621 Build 22621
Arch     x86_64

This project is written in python: executable='D:\Infrastructure\venv\Scripts\python.exe' version='3.8.10
'

Current Stack: auto

...

Found no pending operations associated with auto

Backend
Name           pulumi.com
URL            https://app.pulumi.com/xxxxxxxxxx
User           xxxxxxxxxx
Organizations  xxxxxxxxxx

Dependencies:
NAME                 VERSION
azure-storage-blob   12.14.1
pem                  21.2.0
pip                  21.3.1
pip-chill            1.0.1
pulumi-azure-native  1.67.0
pyOpenSSL            23.0.0
setuptools           60.2.0
wheel                0.37.1

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by abhinav** [comment]
Thanks for filing the issue, @msamsami.
I agree that it's desirable for dynamic providers to support async method implementations.
Although I'm marking this as a feature request rather than a bug because this is a request to support functionality that is not currently supported.

**Comment by msamsami** [comment]
> Thanks for filing the issue, @msamsami.
> I agree that it's desirable for dynamic providers to support async method implementations.
> Although I'm marking this as a feature request rather than a bug because this is a request to support functionality that is not currently supported.

Thanks @abhinav. However, it is mentioned in [this web page](https://www.pulumi.com/docs/concepts/resources/dynamic-providers/) that dynamic providers support async methods. This is why I thought my implementation should work. 

```
Each of these methods can be asynchronous, and most implementations of these methods will perform network I/O to provision resources in a backing cloud provider or other resource model. 
```

**Comment by Frassle** [comment]
There's some code in the provider to deal with some async call it does by doing:
```
loop = asyncio.new_event_loop()
result = loop.run_until_complete(task)
loop.close()
```
You might be able to wrap your async methods with similar.

**Comment by msamsami** [comment]
I used the following workaround for now (using `asyncio.run`):

```python
class AccessPolicyProvider(ResourceProvider, _AzureRESTAPI):

    async def create_async(self, args) -> CreateResult:
        tenant_id, subscription_id, access_token = await self.get_access_token()
        ....
        return CreateResult("keyvault-access-policy-" + binascii.b2a_hex(os.urandom(8)).decode("utf-8"), outs=args)

    def create(self, args) -> CreateResult:
        return asyncio.run(self.create_async(args))

    async def delete_async(self, id, args):
        ...

    def delete(self, id, args):
        return asyncio.run(self.delete_async(id, args))
    
    async def update_async(self, id, old_inputs, new_inputs) -> UpdateResult:
        ...

    def update(self, id, old_inputs, new_inputs) -> UpdateResult:
        return asyncio.run(self.update_async(id, old_inputs, new_inputs))
```

**Comment by dlecocq** [comment]
I'm also quite interested in this. It seems like the lack of this support would be an impediment to being able to provision huge numbers of resources - or maybe there's an underlying mechanism that I'm not grokking?
